tests/test_transforms.py

```diff
@@ -1428,3 +1429,37 @@ def test_randomgridshuffle():
         assert len(w) == 1
         assert w[0].category is UserWarning
     warnings.resetwarnings()
+
+
+@pytest.mark.parametrize("num_shadows_limit, num_shadows_lower, num_shadows_upper, expected_warning", [
+    ((1, 2), None, None, None),
+    ((2, 3), None, None, None),
+    ((1, 2), 1, None, DeprecationWarning),
+    ((1, 2), None, 2, DeprecationWarning),
+    ((1, 2), 1, 2, DeprecationWarning),
+    ((2, 1), None, None, ValueError),
+])
+def test_deprecation_warnings_random_shadow(
+    num_shadows_limit: Tuple[int, int],
+    num_shadows_lower: Optional[int],
+    num_shadows_upper: Optional[int],
+    expected_warning: Optional[Type[Warning]],
+) -> None:
+    """
+    Test deprecation warnings for RandomShadow
+    """
+    with warnings.catch_warnings(record=True) as w:
+        warnings.simplefilter("always")
+        if expected_warning == ValueError:
```

Comment:
<a href="https://sourcery.ai"><img src="https://sourcery.ai/favicon.svg" height="15" alt="Sourcery logo" /></a>  **suggestion (testing):** Ensure test coverage for the shadow generation logic in `RandomShadow`.

The current tests focus on parameter validation and deprecation warnings. It would be beneficial to extend the test coverage to include the shadow generation logic itself, ensuring that the shadows are generated within the specified ROI and with the expected number of shadows.